2 lab(simple_mvvm_app): реализовал лабораторную 2 по предмету ООАиП.

RockInHead opened the pull request:
Реализовал лабу №2.
Программа сохраняет значение текстовых в json файл, а также умеет считывать из  json файла. Также имеет адаптивную верстку.

RockInHead committed:
Инициализационный коммит.
Приступил к 2ой лабе. Создал новое решение Contacts типа WPF Application. Название csproj переименовал в View.

RockInHead committed:
Создал 2 папки: Model и ViewModel. В папке Model создал класс Contact, хранящий имя, номер телефон и почту контакта. С автосвойствами и конструктором. Начальный вариант, в дальнейшем класс будет доработан.

RockInHead committed:
Промежуточный коммит.
Начал верстку графического приложения.

RockInHead committed:
Реализовал графический интерфейс с иконкой.

RockInHead committed:
Смог связать View c VM и VM с Model.
Требуется рефакторинг.

RockInHead committed:
Реализовал сохранение и выгрузку в json.
Сделал комментарии ко всему коду, сделал рефакторинг.

sandra155464 commented on Programming/Contacts/Contacts/App.xaml:
```diff
@@ -0,0 +1,9 @@
+﻿<Application x:Class="Contacts.App"
+             xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
+             xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
+             xmlns:local="clr-namespace:Contacts"
```
Убрать неиспользуемое пространство имён (local)

sandra155464 commented on Programming/Contacts/Contacts/App.xaml.cs:
```diff
@@ -0,0 +1,14 @@
+﻿using System.Configuration;
+using System.Data;
```
Удалить неиспользуемые директивы using

sandra155464 commented on Programming/Contacts/Contacts/App.xaml.cs:
```diff
@@ -0,0 +1,14 @@
+namespace Contacts
+{
+    /// <summary>
+    /// Interaction logic for App.xaml
```
Добавить точку в конце комментария

sandra155464 commented on Programming/Contacts/Contacts/MainWindow.xaml:
```diff
@@ -0,0 +1,74 @@
+        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
+        xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
+        xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
+        xmlns:local="clr-namespace:Contacts"
```
Также убрать неиспользуемое пространство имён local

sandra155464 commented on Programming/Contacts/Contacts/MainWindow.xaml:
```diff
@@ -0,0 +1,74 @@
+        Icon="contacts_96x96.ico"
+        Title="Contacts" Height="375" Width="500" MinWidth="500" MinHeight="300" Padding="15">
+        
+    <Grid Margin="15,15,15,15">
```
Margin="15,15,15,15" - это то же самое, что Margin="15". Также у XAML-тэгов атрибуты, если их больше 2, нужно писать каждый на отдельной строке, а ещё правильно расставлять табуляцию. Чтобы в этот и следующий разы не заниматься форматированием вручную, поставь себе расширение XAML Styler и настрой его следующим образом:
![image](https://github.com/user-attachments/assets/6da82c89-4d63-4352-ba89-bb9f8eb9856b)

sandra155464 replied:
Замечание про форматирование XAML-файлов не исправлено. Если у тебя нормально установился XAML Styler, то у тебя должны сработать горячие клавиши Ctrl+K, Ctrl+2 и содержимое файла отформатируется

sandra155464 replied:
Замечание не исправлено - проверь порядок в классах MainVM, LoadCommand, Contact

RockInHead replied:
Прости, не заметил новый комментарий)
Исправлю)

sandra155464 commented on Programming/Contacts/Contacts/Model/Contact.cs:
```diff
@@ -0,0 +1,46 @@
+public class Contact
+{
+    /// <summary>
+    /// Задает и возвращет имя контакта.
```
Чтобы в дальнейшем видеть опечатки и не отправлять код с неправильными комментариями, поставь себе также расширение SpellChecker. Перед отправкой коммита с исправлениями проверь ещё раз все комментарии на опечатки )

RockInHead replied:
Скачал SpellChecker, но к сожалению не смог разобраться, как настроить на русский язык(
Поэтому пробежался глазами по комменатриям.

sandra155464 replied:
На Яндекс.Диске есть папка ГПО/Статьи по процессу разработки, и в ней есть инструкция с настройками инструментов, среди которых SpellChecker. Почитай и попробуй ещё раз настроить

sandra155464 commented on Programming/Contacts/Contacts/Model/Services/ContactSerializer.cs:
```diff
@@ -0,0 +1,60 @@
+    /// Инициализирует новый экземпляр <see cref="ContactSerializer"/>.
+    /// Устанавливает путь к файлу по умолчанию и создаёт каталог, если он отсутствует.
+    /// </summary>
+    public ContactSerializer()
```
Мне кажется, что по логике приложения тебе не понадобится несколько экземпляров сериалайзера, поэтому из этого класса вполне можно убрать конструктор и сделать его статическим. Всю логику из конструктора можно вынести в отдельный метод, например, CreateDirectory(), и вызывать его непосредственно перед сохранением контакта

sandra155464 commented on Programming/Contacts/Contacts/ViewModel/LoadCommand.cs:
```diff
@@ -0,0 +1,57 @@
+    /// <summary>
+    /// Событие, вызываемое при изменении состояния выполнения команды.
+    /// </summary>
+    public event EventHandler CanExecuteChanged;
```
Событие CanExecuteChanged у тебя в самом конце класса - это неправильный порядок расположения. Члены класса должны располагаться в следующем порядке:

1. Вложенные перечисления
2. Вложенные классы
3. Константы
4. Статические поля
5. Поля доступные только для чтения (readonly)
6. Перечисления
7. Внутренние классы
8. Поля
9. Конструкторы
10. Делегаты
11. События
12. Свойства
13. Публичные методы
14. Защищенные методы
15. Приватные методы

Поправь в этом и во всех других классах, если там тоже будет неправильно )

sandra155464 replied:
Замечание не исправлено - проверь порядок в классах MainVM, LoadCommand и Contact

sandra155464 commented on Programming/Contacts/Contacts/ViewModel/MainVM.cs:
```diff
@@ -0,0 +1,121 @@
+public class MainVM : INotifyPropertyChanged
+{
+    /// <summary>
+    /// Объект текущего контакта.
```
Если поле _contact хранит объект _текущего_ контакта, то надо переименовать его на _currentContact

sandra155464 replied:
Свойство, возвращающее текущий контакт, также должно быть названо CurrentContact

sandra155464 commented on Programming/Contacts/Contacts/ViewModel/MainVM.cs:
```diff
@@ -0,0 +1,121 @@
+﻿using System.ComponentModel;
+
+/// <summary>
+/// Основная ViewModel для управления контактами и их сохранением/загрузкой.
```
"ViewModel" лучше заменить на "модель представления". Если в XML-комментарии ты не используешь TODO или тэг cref, то лучше писать по-русски )

sandra155464 commented on Programming/Contacts/Contacts/ViewModel/MainVM.cs:
```diff
@@ -0,0 +1,121 @@
+    /// <summary>
+    /// Объект для сериализации и десериализации контактов.
+    /// </summary>
+    private readonly ContactSerializer _serializer;
```
Поле _serializer только передаётся в качестве параметра для конструкторов классов SaveCommand и LoadCommand. Его можно сразу поместить в эти классы и в них же инициализировать, а из MainVM убрать

sandra155464 commented on Programming/Contacts/Contacts/ViewModel/MainVM.cs:
```diff
@@ -0,0 +1,121 @@
+    private readonly ContactSerializer _serializer;
+
+    /// <summary>
+    /// Текущий контакт.
```
Поправь комментарии к свойствам, переписав их аналогично комментариям к свойствам в классе Contact.cs, через "возвращает и задаёт"

sandra155464 commented on Programming/Contacts/Contacts/ViewModel/MainVM.cs:
```diff
@@ -0,0 +1,121 @@
+    /// <param name="contact">Загруженный контакт.</param>
+    private void UpdateContact(Contact contact)
+    {
+        if (contact != null)
```
Чтобы повысить читаемость, можно переписать логику метода так:
```cs
if (contact == null)
{
    return;
}

Contact = contact;
OnPropertyChanged(nameof(Name));
OnPropertyChanged(nameof(PhoneNumber));
OnPropertyChanged(nameof(Email));
``` 
Исправь по аналогии в других местах в коде, где у тебя есть блоки if.

Кроме того, можно изменить сеттер свойства Contact следующим образом, чтобы по аналогии с другими свойствами у тебя при присваивании тут же зажигались события обновления значений:
``` cs
set
{
    _contact = value;
    OnPropertyChanged(nameof(Name));
    OnPropertyChanged(nameof(PhoneNumber));
    OnPropertyChanged(nameof(Email));
}
```

sandra155464 replied:
Вчера не заметила, но во всех сеттерах у тебя отсутствует проверка на то, что присваиваемое значение отличается от текущего. Например, у свойства Contact эта проверка будет выглядеть так:
```cs
if (_contact == value)
{
    return;
}

_contact = value;
OnPropertyChanged(nameof(Name));
OnPropertyChanged(nameof(PhoneNumber));
OnPropertyChanged(nameof(Email));
```

Во всех остальных сеттерах других классов нужно реализовать такую же проверку

sandra155464 commented on Programming/Contacts/Contacts/ViewModel/SaveCommand.cs:
```diff
@@ -0,0 +1,58 @@
+/// </summary>
+public class SaveCommand : ICommand
+{
+
```
Убрать пустую строку. 56 строка тоже пустая, её тоже стоит удалить

sandra155464 commented:
Оставила замечания по коду, и есть ещё одно небольшое замечание по именованию коммитов и пул-реквестов. Прикреплю ниже файл со стандартом именования - в следующий раз при отправке коммита или создании пула сверяйся по нему )
[Git__-_Overview.pdf](https://github.com/user-attachments/files/18920635/Git__-_Overview.pdf)

RockInHead committed:
style(App.xaml): удалил неиспользуемое пространство имён (local).

- учусь писать правильные коммиты

RockInHead committed:
style(App.xaml.css): убрал лишние using и оформелние комментария

- удалены неиспользуемые usings
- добавлена точка в конце xml комментария

RockInHead committed:
style(MainWindow.xaml): удалил лишний using и поправил Margin.

- скачен и настроен XAML Styler
- удалены лишние параметры у Margin
- удален лишний using (local)

RockInHead committed:
refactoring(ContactSerializer): сделал класс статическим.

- убраны лишнее создание экземпляров класс ContactSerializer
- изменены конуструкторы для SaveCommand и LoadCommand
- Изменены xml-комментарии для выше описанных классов

RockInHead committed:
style(LoadCommand, SaveCommand): перенес элементы в правильном порядке.

- перенесён CanExecuteChanged после конструктора в LoadCommand.cs
- перенесён CanExecuteChanged после конструктора в SaveCommand.cs

RockInHead committed:
refactoring(MainVM): изменил функцию UpdateContact.

- переименована переменная _contact в _currentContact
- изменены сеттеры в свойствах на проверку текущего и входящего значений
- изменен ModelView на Модель представления в xml-комментарияx
- изменена функция UpdateContact, теперь более эффектинвная

RockInHead committed:
style(View): удалил лишние using, прверил xml-комментарии.

sandra155464 commented on Programming/Contacts/Contacts/Model/Services/ContactSerializer.cs:
```diff
@@ -2,20 +2,19 @@
+/// Статический класс для сериализации и десериализации контактов в JSON-файл.
 /// </summary>
-public class ContactSerializer
+static public class ContactSerializer
```
Во всём классе поменять местами static и public/private

sandra155464 commented on Programming/Contacts/Contacts/ViewModel/MainVM.cs:
```diff
@@ -1,68 +1,85 @@
-    /// Текущий контакт.
+    /// Возвращает и задает текущий контакт.
     /// </summary>
     public Contact Contact
```
Свойство, возвращающее и задающее текущий контакт, также должно быть названо CurrentContact

sandra155464 commented on Programming/Contacts/Contacts/ViewModel/SaveCommand.cs:
```diff
@@ -5,30 +5,28 @@
+    public SaveCommand(Func<Contact> getContact)
     {
-        _serializer = serializer ?? throw new ArgumentNullException(nameof(serializer));
+        ContactSerializer.CreateDirectory();
```
Этот метод лучше перенести в метод SaveContact() класса ContactSerializer, т.к. классу команды не обязательно управлять созданием директории

sandra155464 commented on Programming/Contacts/Contacts/ViewModel/MainVM.cs:
```diff
@@ -1,68 +1,85 @@
 
     /// <summary>
-    /// Свойство для email контакта с уведомлением об изменении.
+    /// Возвращает и задает текущцю почту контакта.
```
Опечатка

sandra155464 commented:
Закрыла часть старых замечаний, оставила новые

sandra155464 commented on Programming/Contacts/Contacts/Model/Services/ContactSerializer.cs:
```diff
@@ -0,0 +1,59 @@
+    static public void CreateDirectory()
+    {
+        _filePath = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.MyDocuments), "Contacts", "contacts.json");
+        string directory = Path.GetDirectoryName(_filePath);
```
В прошлый раз забыла написать - вместо string нужно в таких случаях использовать var. Исправь в этом методе и в методах ниже

RockInHead committed:
refactoring(ContactSerializer): заменил string на var и поменял местами static и public/private.

RockInHead committed:
Revert "refactoring(ContactSerializer): заменил string на var и поменял местами static и public/private."

This reverts commit 58d4a2a78afb1e64cd276defb7959f411cacdd10.

RockInHead committed:
refactoring(ContactSerializer): заменил string на var и поменял местами static и public/private."

RockInHead committed:
style(MainVM): поправил опечатку в xml комментарии и поменял имя свойства.

RockInHead committed:
refactoring(SaveCommand,ContactSerializer): вынес создания новой папки из SaveCommand в ContactSerializer.

RockInHead commented:
Закрыл замечания.
Можешь пожалуйста еще дать обратную связь насчет моей выполненной лабораторной и именования коммитов? 🥺

sandra155464 commented:
Новые замечания закрыла, у тебя остались лишь старые замечания про XAML Styler, Spell Checker и порядок расположения классов - пока не вижу, что ты их поправил, поэтому на всякий случай написала о них тут, если ты вдруг забыл.
Если до четверга не сможешь разобраться со Spell Checker'ом или XAML Styler'ом (или ещё с чем-то), то после собрания я попробую помочь тебе их настроить. До четверга тоже можешь задавать вопросы (возможно, получится удалённо помочь тебе настроить всё), только, пожалуйста, помни про время - после 23.00 лучше никого не тэгать, потому что это уже ооочень поздно.
Насчёт обратной связи:
1) по лабораторной: на мой взгляд, технически всё выполнено неплохо - программа выполняет то, что указано в методичке, аварийные ситуации (например, отсутствие директории для файла с сериализованными данными) учтены, и программа при них не ломается. Отдельно для себя я отметила способ передачи экземпляра контакта в SaveCommand и получения его из LoadCommand - использование делегатов выглядит интересно, я в своё время делала по-другому )
Единственный (на данный момент, как минимум) момент - это изначально неправильное расположение поля с экземпляром сериалайзера. По своему опыту скажу, что подобного рода ошибки начинаешь замечать только с опытом, поэтому пока что нормально )
А вот по оформлению кода, как ты мог заметить, замечаний много. И хоть эти замечания иногда встречаются и у хоть сколько-нибудь опытных разработчиков, но лучше бы, чтобы их было поменьше. Поэтому когда закроешь эту лабу и приступишь к следующей, внимательно просматривай свой код и поглядывай на замечания в своих старых пулах - возможно, некоторые ошибки у тебя будут дублироваться, и ты сможешь исправить их до ревью, тем самым ускорив процесс проверки.
2) по именованию коммитов: до первого моего ревью именование было неправильным, но оно и понятно. После того, как я тебе скинула файл с правилами, с именованием всё стало хорошо, проблем не вижу :)

RockInHead committed:
style(MainWindow.xaml): Разобрался с XAML Stylerом, исправил форматирование.

RockInHead committed:
style(View): Вроде разобрался с Spell Checker и смог настроить его на русский язык. Поправил опечатки.

RockInHead committed:
style(View): Расположил элементы в классах в правильном утвержденном порядке.

Утвержденный порядок элементов в классе:
1. Вложенные перечисления
2. Вложенные классы
3. Константы
4. Статические поля
5. Поля доступные только для чтения (readonly)
6. Перечисления
7. Внутренние классы
8. Поля
9. Конструкторы
10. Делегаты
11. События
12. Свойства
13. Публичные методы
14. Защищенные методы
15. Приватные методы

sandra155464 commented on Programming/Contacts/Contacts/Model/Services/ContactSerializer.cs:
```diff
@@ -4,20 +4,20 @@
-    static public void CreateDirectory()
+    public static void CreateDirectory()
     {
         _filePath = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.MyDocuments), "Contacts", "contacts.json");
```
Длина строки превышает 100 символов. Перенеси на отдельные строки Environment.SpecialFolder.MyDocuments), "Contacts" и  "contacts.json"

sandra155464 commented on Programming/Contacts/Contacts/Model/Services/ContactSerializer.cs:
```diff
@@ -29,29 +29,29 @@ static public void CreateDirectory()
         }
-
-        string json = JsonConvert.SerializeObject(contact, Formatting.Indented);
+        CreateDirectory();
```
После фигурной скобочки нужна одна пустая строка

sandra155464 commented on Programming/Contacts/Contacts/Model/Services/ContactSerializer.cs:
```diff
@@ -4,20 +4,20 @@
         _filePath = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.MyDocuments), "Contacts", "contacts.json");
-        string directory = Path.GetDirectoryName(_filePath);
+        var directory = Path.GetDirectoryName(_filePath);
 
```
А вот тут убрать пустую строку

RockInHead committed:
style(ContactSerializer): закрыл оставшиеся замечания, убрал лишние строки, сделал код более читаемым.
